Create initial sharding playbook (#254)
sigstore/rekor-tiles: SHARDING.md

# Sharding Playbook

This playbook walks through the steps to create a new log shard.
The steps should be the same for Rekor and the CT log, though the

> CT process for generating key material may differ.

Reviewer: rekor shouldn't know about CT, or this should be in another repo?

Author: Will update, just wanted to state that the Rekor and CT sharding processes should be the same.

Roughly, the steps are:

> 1. Create the GCP resources with Terraform

Reviewer: same here: this seems specific to our public good instance.

Author: Yep, I'll move this playbook to PGI.

2. Generate key material
3. Spin up the new infrastructure
6. Update prober
4. Perform manual testing
5. Update TrustedRoot with the new key material and SigningConfig with the new shard URL
6. Turning down the old shard

## Create GCP resources

New shards are declared as additional modules under the respective
environment. This is a different pattern than other modules, which are all

> initialized through the `sigstore` module. Declaring each shard as its

Reviewer: I'm not sure exactly what this means (probably because I have no context into infra). Why are other modules initialized through the single `sigstore` module?

own module allows us to make updates to shards one-by-one. Otherwise,
by updating the module reference of `sigstore`, all shards would be
concurrently updated.

To add a new shard, create a module under either
`terraform/environments/staging/1-infrastructure/staging.tf` or

> `terraform/environments/production/1-infrastructure/production.tf`.

Reviewer: can these be links to somewhere?

> Name the module `rekor-tiles-log-<year>-<revision>`, e.g.

Reviewer: what does revision mean here? That we can have multiple shards per year? Should they always count up?

nvm, somewhat answered in the example below, but some guidance on revision would still be helpful.

`rekor-tiles-log-2025-1`.

<!--- TODO: Add code pointers for module example -->

Update `general.auto.tfvars` to include the shard variable values:

```
rekor_tiles_shards = {
  "log2025-1" = {
    ...
  },
  "log2025-2" = {
    ...
  },
  "log2026-1" = {
    ...
  }
}
```

After merging, run `terraform plan` through the GitHub Actions workflow
for [staging](https://github.com/sigstore/public-good-instance/actions/workflows/env-staging.yml)
or [prod](https://github.com/sigstore/public-good-instance/actions/workflows/env-prod.yml).
Confirm the expected resources are created, then run `terraform apply` through the workflow.

## Creating Key Material

The public-good instance uses a Tink keyset to sign checkpoints. Tink
encrypts a signing key using a "key encryption key" managed by GCP KMS,
and loads the key into memory on service startup.

Tink is beneficial for reliability and cost reduction - KMS is now only
a startup dependency rather than a runtime dependency, and signing happens
exclusively in memory.

To create a shard signing key:

> 1. Grant yourself access to use the new signing key. Update the `staging.tf`

Reviewer: you can just do this without someone approving it?

  role        = "roles/cloudkms.cryptoOperator"

  member = "user:<username>@sigstore.dev"
}
```

2. Run `terraform plan` and `terraform apply` using the GHA workflow.

3. Using [tinkey](https://developers.google.com/tink/tinkey-overview#installation), create an encrypted Tink keyset:

```
tinkey create-keyset --key-template ED25519 --out enc-keyset.cfg --master-key-uri gcp-kms://projects/<project name>/locations/<region>/keyRings/<key-ring>/cryptoKeys/<key>
```

Save the output file `enc-keyset.cfg`, which will be used as a value in the Helm chart.

4. Revoke IAM access by removing `google_kms_key_ring_iam_member`, and run `plan` and `apply`.

## Creating K8s Resources

Update `argocd/bootstrap-utilities/staging/values.yaml` or `argocd/bootstrap-utilities/production/values.yaml`
to include the new shard. Under `rekor-tiles.shards`, copy the previous shard instantiation, and update
the values accordingly based on the Terraform resources names. For `keyset`, copy the contents of
`enc-keyset.cfg` (Since this value is encrypted, it's secure to include it in the config rather than
use GCP Secrets).

<!--- TODO: Add code pointers for values -->

> Merge, and wait for ArgoCD to create the resources and spin up the service.

Reviewer: wait how long? is there some indicator? maybe an an oncall person would already know this?

## Update Prober

> Update the prober configuration for

Reviewer: does prober work with v2?

Author: Tracking in https://github.com/sigstore/rekor-tiles/issues/46

[staging](https://github.com/sigstore/sigstore-probers/blob/8608ff2cbc6a9121f12c75cb01a18f32e1428195/.github/workflows/prober-staging.yml#L28)
or
[prod](https://github.com/sigstore/sigstore-probers/blob/8608ff2cbc6a9121f12c75cb01a18f32e1428195/.github/workflows/prober-prod.yml#L26)
to specify the new shard name. Merge and kick off the workflow.

<!--- TODO: Update prober to support testing against multiple Rekor shard URLs -->
<!--- TODO: Update prober to test reads and writes separately, since frozen shards will still support reads -->

## Perform Manual Testing

Using Cosign, test signing and verification against this new shard. You will need to manually update the SigningConfig
and TrustedRoot files and pass them via CLI flags since those files are pulled from TUF and will not yet have been updated.

> <!--- TODO: Update with example TrustedRoot and SigningConfig -->

Reviewer: include example call to cosign using custom trustedroot and signingconfig.

Author: Updated with a curl-based example rather than cosign.

## Update TUF Verification Material

Now that the shard is up and we've successfully signed and verified using it, we will
update the TrustedRoot and SigningConfig files distributed via TUF so that clients
can verify proofs from the shard and use the shard during signing.

For TrustedRoot, the validity window for the shard should be from the point
when the shard was spun up. For SigningConfig, the validity window should start
after the TUF timestamp expires to guarantee all clients will have picked up
the latest TrustedRoot. In practice for our TUF root with timestamp validity
of one week, this means setting the validity window to be at least one week
past when the TUF targets are signed.

> You will need to estimate two timestamps:

Reviewer: I think we should have a well defined overlap period. maybe 1 day/week/month or whatever?

Having to guess this means people will ask around to get a recommendation instead of us just having this decision already made.

> * New shard's start for SigningConfig, which should be one week past when the TUF
> timestamp will be signed. Overestimate if needed, as you don't want clients to start

Reviewer 2: Would this be more understandable to folks with no knowledge of root-signing details?

```suggestion
* New shard's start for SigningConfig, which should be one week past when the root-signing PR is merged and new TUF repository is published. Overestimate if needed, as you don't want clients to start
```

using the new shard before all clients have picked up the latest TrustedRoot.

> * Previous shard's end for TrustedRoot, which is when the previous shard's write path
> is turned off. This must be after the start of the new shard. It doesn't have to

Reviewer 2: ```suggestion
* Previous shard's end for TrustedRoot: clients will consider this shards entry timestamps invalid if they are after this date. This must be after the start of the new shard. It doesn't have to
```

I don't know if the suggestion is any better -- you could reword the original as well to make it a bit clearer (that this date doesn't need to be the moment shard is turned off, just needs to be later than the turn off date.

overly precise, and can be updated in a future TUF signing event.

Update the TrustedRoot for the new shard's key material:

```
{
  "mediaType": "...",
  "tlogs": [
    // new shard
    {
      "baseUrl": "https://<year-revions>.rekor.(sigstore|sigstage).dev",
      "hashAlgorithm": "SHA2_256",
      "publicKey": {
        "rawBytes": "base64-encoded DER public key",
        "keyDetails": "ED25199",
        "validFor": {
          "start": "<UTC timestamp for when shard was spun up>"
        }
      },
      "logId": {
        "keyId": "wNI9atQGlz+VWfO6LRygH4QUfY/8W4RFwiT5i5WRgB0="
      }
    },
    // previous shard
    {
      "baseUrl": "...",
      "hashAlgorithm": "SHA2_256",
      "publicKey": {
        "rawBytes": "...",
        "keyDetails": "...",
        "validFor": {
          "start": "...",

>           "end": "<UTC timestamp for approx when the shard will be shut down>"

Reviewer 2: ```suggestion
          "end": "<UTC timestamp that is after the shard is turned read only>"
```

        }
    ],
    "rekor_tlog_config": {...},
    "tsa_urls": [...],
    "tsa_config": {...},
}
```

Follow the playbooks in https://github.com/sigstore/root-signing and
https://github.com/sigstore/root-signing-staging to orchestrate a new
signing event.

## Turn Down Old Shard

Wait a week, until all clients have received updated SigningConfigs
and started signing using the new shard, and are using the updated TrustedRoot
to verify inclusion proofs.

Verify write traffic to the previous shard has gone down to ~0 QPS.

<!--- TODO: Update with link to monitoring -->

### Prober

Update the prober configuration to remove the previous shard URL for testing writes.

### K8s

Turn down the previous shard by removing the previous values for `rekor-tiles.shards`.
We can turn down the Kubernetes pods because the server only serves the write path -
the read path is exclusively served via GCP resources (Load Balancer, Storage).

Merge and wait for ArgoCD to shut down the server.

### Remove GCP Resources

We only need a subset of GCP resources to serve read traffic. To save costs,
we can turn down the database and destroy the signing key.

<!--- TODO: Specify how we should do this -->

> <!--- TODO: Specify how to convey to monitors the shard is frozen.

Author: One other thought - For CT, we can't add extension lines, so we if we want to denote a frozen shard in the same way, baking in the checkpoint might be the only viable approach.

4. Verify the health of the infrastructure with manual testing
5. Run prober tests
6. Update TrustedRoot with the new key material and SigningConfig with the new shard URL
7. Turn down the old shard

## Terraform Module Changes

New shards are declared as additional modules under the respective
environment. This is a different pattern than other modules, which are all
initialized through the
[`sigstore` module](https://github.com/sigstore/terraform-modules/tree/main/gcp/modules/sigstore).
This design allowed deployers to initialize this one module and set all required variables, at
which point you'd have a complete Sigstore stack on GCP. The downside of this approach is
private deployers that don't need all GCP infrastructure have to initialize all resources,
though deployers can always initialize each module referenced in `sigstore` individually.

The other downside to referencing the `sigstore` module is that an update to a sub-module
will concurrently update all deployed shards.  Declaring each shard as its own module allows us
to make updates to shards one-by-one.

The tiled transparency log module we'll reference in Terraform is

> [`tiles_tlog`](https://github.com/sigstore/terraform-modules/tree/main/gcp/modules/tiles_tlog).

Author: This is just to give background, either seems fine.

This means that every backend connection from the frontend load balancer to the Compute VM is functional, as verified
by a healthcheck for HTTP and HTTP/2 traffic.

Under Backend buckets, you should see a storage bucket and Cloud CDN should be Enabled.

### Verify the read path

Verify that a checkpoint is served from the Storage bucket, e.g.

```
curl https://log2026-1.rekor.sigstage.dev/api/v2/checkpoint

log2026-1.rekor.sigstage.dev
0
YR3vjr9qTpYPib0tUJT/7/pAPuK9xDOibMncp2aQAJE=

— log2026-1.rekor.sigstage.dev 8w1amdxKovmc7H9+4lnx/vSeAQo9zo8cr8EHFCVviWMtTmm/Xw+zGDnBni9/mHHa1hFXIQlC29+YyaOXRxduvbE20Qg=
```

### Verify the write path

Following
[these instructions](https://github.com/sigstore/rekor-tiles/blob/main/CLIENTS.md#rekor-v2-the-bash-way),
test the write path for the new shard. This will sign an entry with a private key and upload
the signing event to the log.

## WIP: Update Prober

Tracked in https://github.com/sigstore/rekor-tiles/issues/46. We're aiming to avoid
needing to manually update the prober. The prober should pick up new shards based on the
SigningConfig (for the write path) and TrustedRoot (for the read path).

> We will need a mechanism for running prober tests against a new deployment before we've

Reviewer 3: Also, v2 shards aren't directly discoverable from the trusted root, because the trusted root does not contain `apiMajorVersion`, like the signing config does. We can work around this by querying `<shard url>/api/v2/checkkpoint`, though.

> We will need a mechanism for running prober tests against a new deployment before we've

Author: I updated this to specify that you should use the SigningConfig for shard discovery, as it will contain both the active and all inactive shards. If a prober needs to correlate between the SigningConfig and TrustedRoot, it can use the shard URL.

            }
        },
        // previous shard
        {
            ...,
            "valid_for": {
                "start": "...",
                "end": "<valid_for.start for new shard>"
            }
        }
    ],
    "rekor_tlog_config": {...},
    "tsa_urls": [...],
    "tsa_config": {...},
}
```

Follow the playbooks in https://github.com/sigstore/root-signing and
https://github.com/sigstore/root-signing-staging to orchestrate a new
signing event.

## Turn Down Old Shard

Wait a week, until all clients have received updated SigningConfigs
and started signing using the new shard, and are using the updated TrustedRoot
to verify inclusion proofs.

### Verify write traffic

Verify write traffic to the previous shard has gone down to ~0 QPS.
If there is significant traffic, look at the user agent and see if you
can figure out who's calling the service.

<!--- TODO: Update with link to monitoring -->

### Post on Slack

Make a post on Slack letting the community know we'll be freezing
and turning down the log.

### Disable any alerts

Make sure to either disable or remove via Terraform any alerts
for failing healthchecks or missing metrics. As we turn down the service

> and delete resources, we expect that healthchecks will start to fail.

Author: Changed this section to a WIP since we don't have the list of alerts yet.

Update `production.tf` or `staging.tf` based on the example. Make sure to update
the `module` name, `shard_name` variable, and the name of the map key for `var.rekor_tiles_shards`.

After merging, run `terraform plan` through the GitHub Actions workflow
for [staging](https://github.com/sigstore/public-good-instance/actions/workflows/env-staging.yml)
or [prod](https://github.com/sigstore/public-good-instance/actions/workflows/env-prod.yml).
Confirm the expected resources are created, then run `terraform apply` through the workflow.
Note this will also remove the org restriction.

### Create Key Material

The public-good instance uses a Tink keyset to sign checkpoints. Tink
encrypts a signing key using a "key encryption key" (KEK) managed by GCP KMS,
and loads the key into memory on service startup.

Tink is beneficial for reliability and cost reduction - KMS is now only
a startup dependency rather than a runtime dependency, and signing happens
exclusively in memory.

To create a shard KEK:

1. Grant yourself access to use the new KEK. Update the `staging.tf`
or `production.tf` config to include:

```
resource "google_kms_key_ring_iam_member" "rekor-tiles-keyring" {

>   key_ring_id = "<project-rekor or projectsigstore-staging>/<location, e.g. global>/<name of shard keyring>"

Author: Thanks, I added how the string is constructed.

  role        = "roles/cloudkms.cryptoOperator"

  member = "user:<username>@sigstore.dev"
}
```

Request a review from an infrastructure maintainer and merge.

[Example PR](https://github.com/sigstore/public-good-instance/pull/2942)

2. Run `terraform plan` and `terraform apply` using the GHA workflow.

3. [Install the gcloud CLI](https://cloud.google.com/sdk/docs/install) if not already installed.

4. Authenticate to gcloud with application credentials: `gcloud auth application-default login`

5. Using [tinkey](https://developers.google.com/tink/tinkey-overview#installation), create an encrypted Tink keyset:

```

> tinkey create-keyset --key-template ED25519 --out enc-keyset.cfg --master-key-uri gcp-kms://projects/<project name>/locations/<region>/keyRings/<key-ring>/cryptoKeys/checkpoint-signer-key-encryption-key

Author: For the 2025-1 shard, [the key name variable](https://github.com/sigstore/public-good-instance/blob/3c97f5b6b6f52d2eb38018619555c57026241a67/terraform/environments/staging/1-infrastructure/staging.tf#L242) is set to `checkpoint-signer`. For future shards, we can just omit the key name variable and the use the new default `checkpoint-signer-key-encryption-key`.

SigningConfig, with the latest active shard having no end date. Only the active
shard will accept write traffic. All other shards in the SigningConfig can still
handle read traffic.

We will need a mechanism for running prober tests against a new deployment before we've
distributed an updated SigningConfig and TrustedRoot. Assuming this is a parameter to
a GitHub Actions workflow, a deployer should provide the URL to the new shard and run
prober tests.

## Update TUF Verification Material

Now that the shard is up and we've successfully signed and verified using it, we will
update the TrustedRoot and SigningConfig files distributed via TUF so that clients
can verify proofs from the shard and use the shard during signing.

For TrustedRoot, the validity window for the shard should be from the point
when the shard was spun up. For SigningConfig, the validity window should start
after the TUF timestamp expires to guarantee all clients will have picked up
the latest TrustedRoot. In practice for our TUF root with timestamp validity
of one week, this means setting the validity window to be at least one week
past when the TUF targets are signed.

> You will need two timestamps:

Author: Thanks, updated.

    "rekor_tlog_config": {...},
    "tsa_urls": [...],
    "tsa_config": {...},
}
```

Follow the playbooks in https://github.com/sigstore/root-signing and
https://github.com/sigstore/root-signing-staging to orchestrate a new
signing event.

## Turn Down Old Shard

Wait a week, until all clients have received updated SigningConfigs
and started signing using the new shard, and are using the updated TrustedRoot
to verify inclusion proofs.

### Verify write traffic

Verify write traffic to the previous shard has gone down to ~0 QPS.
If there is significant traffic, look at the user agent and see if you
can figure out who's calling the service.

<!--- TODO: Update with link to monitoring -->

### Post on Slack

Make a post on Slack letting the community know we'll be freezing
and turning down the log.

### WIP: Disable alerts for old shard

Make sure to either disable or remove via Terraform alerts for the old shard
for failing healthchecks for the write path or missing metrics. As we turn down the service
and delete resources, we expect that healthchecks will start to fail.

As we create alerting, we'll finalize this section with what alerts need to be removed.

### Freeze the log

> Pass `--read-only=true` to the service, which will stop write traffic and

Reviewer 4: I think we could skip this step and just turn off the K8s deployment.

Author: I'm good with that. The only benefit of freezing the log is to gracefully reject write traffic, but unless we're adding in a waiting period for clients to fail and update to the latest shard (which shouldn't be an issue as clients should be using TUF), this doesn't add much.

return an error to the caller. In the previous shard under `rekorTiles.shards`,
add:

```
extraArgs:
  - "--read-only=true"
```

### Verify write traffic again

Write traffic should now be exactly 0 QPS.

### Delete Kubernetes resources

Turn down the previous shard by removing the previous shard from `rekorTiles.shards`.
We can turn down the Kubernetes pods because the server only serves the write path -
the read path is exclusively served via GCP resources (Load Balancer, Storage).

Merge and wait for ArgoCD to shut down the server.

### Delete GCP resources

We only need a subset of GCP resources to serve read traffic. To save costs,
we can turn down the databases and destroy the KEK (indirectly destroying the signing key).

First, remove the protection bit on the databases. In the Terraform configuration,
set `spanner_database_sequencer_deletion_protection` and `spanner_database_antispam_deletion_protection`

> to `true`. Create a PR, merge, `plan`, and `apply`.

Author: Whoops, thanks.

can verify proofs from the shard and use the shard during signing.

For TrustedRoot, the validity window for the shard should be from the point
when the shard was spun up. For SigningConfig, the validity window should start
after the TUF timestamp expires to guarantee all clients will have picked up
the latest TrustedRoot. In practice for our TUF root with timestamp validity
of one week, this means setting the validity window to be at least one week
past when the TUF targets are signed.

You will need two timestamps:

* New shard's start for SigningConfig, which should be one week past when the root-signing PR is merged
and new TUF repository is published. Overestimate (by an hour or even a day) if needed,
as you don't want clients to start using the new shard before all clients
have picked up the latest TrustedRoot.
* Previous shard's end for TrustedRoot: Clients will consider this shard's entry timestamps invalid
if they are after this date. This must be after the start of the new shard. It doesn't have to
overly precise, and can be updated in a future TUF signing event.

### Retrieve shard public key

> Look at the service logs to find the public key. This is the easiest way to retrieve

Reviewer 3: Add a link to the service logs for an existing shard, so we can easily know how to find the logs for the next shard.

Author: The audience for this doc is someone familiar with the infrastructure so they should be familiar with where to retrieve logs for pods.
